src/dd/GateMatrixDefinitions.cpp

```diff
@@ -0,0 +1,223 @@
+                             {0, 0, 0, {cosTheta, -sinTheta}}}};
+}
+
+inline TwoQubitGateMatrix rzxMat(const fp theta) {
+  const auto cosTheta = std::cos(theta / 2.);
+  const auto sinTheta = std::sin(theta / 2.);
+
+  return TwoQubitGateMatrix{{{cosTheta, {0., -sinTheta}, 0, 0},
+                             {std::complex{0., -sinTheta}, cosTheta, 0, 0},
+                             {0, 0, cosTheta, {0., sinTheta}},
+                             {0, 0, {0., sinTheta}, cosTheta}}};
+}
+
+inline TwoQubitGateMatrix xxMinusYYMat(const fp theta, const fp beta = 0.) {
+  const auto cosTheta = std::cos(theta / 2.);
+  const auto sinTheta = std::sin(theta / 2.);
+  const auto cosBeta = std::cos(beta);
+  const auto sinBeta = std::sin(beta);
+
+  return TwoQubitGateMatrix{
+      {{cosTheta, 0, 0, {-sinBeta * sinTheta, -cosBeta * sinTheta}},
+       {0, 1, 0, 0},
+       {0, 0, 1, 0},
+       {std::complex{sinBeta * sinTheta, -cosBeta * sinTheta}, 0, 0,
+        cosTheta}}};
+}
+
+inline TwoQubitGateMatrix xxPlusYYMat(const fp theta, const fp beta = 0.) {
+  const auto cosTheta = std::cos(theta / 2.);
+  const auto sinTheta = std::sin(theta / 2.);
+  const auto cosBeta = std::cos(beta);
+  const auto sinBeta = std::sin(beta);
+
+  return TwoQubitGateMatrix{
+      {{1, 0, 0, 0},
+       {0, cosTheta, {sinBeta * sinTheta, -cosBeta * sinTheta}, 0},
+       {0, {-sinBeta * sinTheta, -cosBeta * sinTheta}, cosTheta, 0},
+       {0, 0, 0, 1}}};
+}
+} // namespace
+
+namespace dd {
+
+GateMatrix opToSingleQubitGateMatrix(const qc::OpType t,
+                                     const std::vector<fp>& params) {
+  switch (t) {
+  case qc::I:
+    return {1, 0, 0, 1};
+  case qc::H:
+    return {SQRT2_2, SQRT2_2, SQRT2_2, -SQRT2_2};
+  case qc::X:
+    return {0, 1, 1, 0};
+  case qc::Y:
+    return {0, {0, -1}, {0, 1}, 0};
+  case qc::Z:
+    return {1, 0, 0, -1};
+  case qc::S:
+    return {1, 0, 0, {0, 1}};
+  case qc::Sdg:
+    return {1, 0, 0, {0, -1}};
+  case qc::T:
+    return {1, 0, 0, {SQRT2_2, SQRT2_2}};
+  case qc::Tdg:
+    return {1, 0, 0, {SQRT2_2, -SQRT2_2}};
+  case qc::SX:
+    return {std::complex{0.5, 0.5}, std::complex{0.5, -0.5},
+            std::complex{0.5, -0.5}, std::complex{0.5, 0.5}};
+  case qc::SXdg:
+    return {std::complex{0.5, -0.5}, std::complex{0.5, 0.5},
+            std::complex{0.5, 0.5}, std::complex{0.5, -0.5}};
+  case qc::V:
+    return {SQRT2_2, {0., -SQRT2_2}, {0., -SQRT2_2}, SQRT2_2};
+  case qc::Vdg:
+    return {SQRT2_2, {0., SQRT2_2}, {0., SQRT2_2}, SQRT2_2};
+  case qc::U:
+    return uMat(params[0], params[1], params[2]);
+  case qc::U2:
+    return u2Mat(params[0], params[1]);
+  case qc::P:
+    return pMat(params[0]);
+  case qc::RX:
+    return rxMat(params[0]);
+  case qc::RY:
+    return ryMat(params[0]);
+  case qc::RZ:
+    return rzMat(params[0]);
+  default:
+    assert(false && "Invalid single-qubit gate type");
+    return {}; // unreachable
```

Comment:
If you really want to mark this as unreachable, then you can use `qc::unreachable()`.
```suggestion
    assert(false && "Invalid single-qubit gate type");
    qc::unreachable();
```
Although that part of the code is technically reachable as the method is public and can easily be passed a wrong `OpType`.
I'd still be fine with adding the unreachable there (and down below).

Reply:
If this code can get called by a library user, would it be better to raise an exception?

Reply:
Most likely, yes.
Could be an `std::invalid_argument` exception.
That way, such errors get a proper error message even in Release mode.